### PR TITLE
cluster: refactor test classes

### PR DESCRIPTION
**Pull request by alissa-tung**

### CODE REVIEW

File: app/src/test/java/io/hstream/testing/BasicTest.java [hunk 1]
```diff
@@ -37,29 +36,49 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Timeout;
 import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.extension.ExtensionContext;
 import org.testcontainers.containers.GenericContainer;
 
-@ExtendWith(BasicExtension.class)
+@ExtendWith(ClusterExtension.class)
```
Comment (Commelina) [inline app/src/test/java/io/hstream/testing/BasicTest.java:54]: Why switch it to the cluster version?

File: app/src/test/java/io/hstream/testing/BasicTest.java [hunk 2]
```diff
@@ -145,132 +187,6 @@ void testDeleteNonExistingSubscriptionShouldFail() {
...
-            hStreamClient, subscription, "test-consumer-new", res, notify2, lock);
-    consumer2.startAsync().awaitRunning();
-    done = notify2.await(10, TimeUnit.SECONDS);
-    Thread.sleep(1000);
-    consumer2.stopAsync().awaitTerminated();
-    Assertions.assertTrue(done);
-    Assertions.assertEquals(records.size(), res.size());
-    Assertions.assertEquals(records, res);
-  }
-
```
Comment (Commelina) [inline app/src/test/java/io/hstream/testing/BasicTest.java]: ~~So are they useless or have better substitutions now?~~ Ok, I know they are moved to `RestartNodeTest.` However, you may find it is not really suitable to classify them in this way. For example. If I write a test which both kills and restarts nodes and runs on a cluster version test environment, where should it be put?

File: app/src/test/java/io/hstream/testing/ClusterKillNodeTest.java [hunk 3]
```diff
@@ -46,18 +68,40 @@ public void teardown() throws Exception {
   }
 
   @Test
-  void listStreamAfterKillNode() {
+  void listStreamAfterKillNodes() {
+    String stream = randStream(hStreamClient);
     hServers.get(0).close();
-    hStreamClient.listStreams();
+    hServers.get(1).close();
+    Assertions.assertEquals(stream, hStreamClient.listStreams().get(0).getStreamName());
```
Comment (Commelina) [inline app/src/test/java/io/hstream/testing/ClusterKillNodeTest.java:78]: ~~Are you sure that `listStreams` guarantees the order (in time) of streams?~~ The test will use a brand-new test environment so it is OK.

File: app/src/test/java/io/hstream/testing/ClusterKillNodeTest.java [hunk 4]
```diff
@@ -46,18 +68,40 @@ public void teardown() throws Exception {
...
+  void killAllShouldFail() throws Exception {
+    for (int i = 0; i < 3; i++) {
+      terminateHServerWithLogs(0, i);
+    }
+    Assertions.assertThrows(
+        Exception.class,
+        () -> {
+          hStreamClient.listStreams();
+        });
   }
```
Comment (Commelina) [inline app/src/test/java/io/hstream/testing/ClusterKillNodeTest.java]: The test name is misleading. It reads like we can not kill all nodes, or something will fail. But you just want to say the client will fail to perform any operation if all nodes are killed.

File: app/src/test/java/io/hstream/testing/ClusterKillNodeTest.java [hunk 5]
```diff
@@ -46,18 +68,40 @@ public void teardown() throws Exception {
...
+    hServers.get(1).close();
+    Assertions.assertEquals(
+        subscription, hStreamClient.listSubscriptions().get(0).getSubscriptionId());
+  }
+
+  @RepeatedTest(5)
+  void write() throws Exception {
     var streamName = TestUtils.randText();
-    System.out.println("hserver cluster size is " + hServers.size());
+    System.out.println("[DEBUG]: HServer cluster size is " + hServers.size());
```
Comment (Commelina) [inline app/src/test/java/io/hstream/testing/ClusterKillNodeTest.java]: I strongly, maybe not now, but still, encourage you to use a logging framework to do such jobs.

File: app/src/test/java/io/hstream/testing/ClusterKillNodeTest.java (hunk 4 above)
Comment (alissa-tung) [reply app/src/test/java/io/hstream/testing/ClusterKillNodeTest.java]: `testListStreamsShouldFailWhenAllNodesAreUnavailable` now

File: app/src/test/java/io/hstream/testing/BasicTest.java (hunk 1 above)
Comment (alissa-tung) [reply app/src/test/java/io/hstream/testing/BasicTest.java:54]: it was suggested that normal operations should run on both cluster and single node, unless there are some tests need to be operate on single node (like restart the only hserver), so the cluster version is now by default

File: app/src/test/java/io/hstream/testing/ClusterKillNodeTest.java (hunk 5 above)
Comment (alissa-tung) [reply app/src/test/java/io/hstream/testing/ClusterKillNodeTest.java]: 056f050

File: app/src/test/java/io/hstream/testing/BasicTest.java (hunk 1 above)
Comment (Commelina) [reply app/src/test/java/io/hstream/testing/BasicTest.java:54]: That's OK.

File: app/src/test/java/io/hstream/testing/SingleNodeTest.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  @AfterEach
+  public void teardown() throws Exception {
+    hStreamClient.close();
+  }
+
+  // -----------------------------------------------------------------------------------------------
+
+  @Disabled("restart may cause test fail, disable")
```
Comment (Commelina) [inline app/src/test/java/io/hstream/testing/SingleNodeTest.java:54]: Do tests with `restartServer` still fail now?

**Comment by Commelina** [comment]
The failed CI is due to a bug in the latest client. Just omit it.

**Pull request merged**